benches/parser.rs

```diff
@@ -0,0 +1,53 @@
+use criterion::{criterion_group, criterion_main, Criterion};
+use std::{fs, path::PathBuf, str::FromStr};
+
+fn benchmark_quil_corpus(c: &mut Criterion) {
+    from_corpus().iter().for_each(|cfg| {
+        c.bench_function(&cfg.name, |b| {
+            b.iter(|| {
+                let _ = quil_rs::Program::from_str(&cfg.program);
+            })
+        });
+    })
+}
+
+struct QuilBenchConfig {
+    name: String,
+    program: String,
+}
+
+fn from_corpus() -> Vec<QuilBenchConfig> {
+    const PATH_SRC: &str = "benches/quilc/tests/good-test-files";
+
+    // collect valid quil programs
+    let mut programs = vec![];
+    let corpus_dir = PathBuf::from(PATH_SRC);
+    let dir = fs::read_dir(corpus_dir).expect("failed to locate quil corpus directory");
+
+    dir.filter_map(Result::ok)
+        .filter(|entry| {
+            entry
+                .metadata()
+                .expect("failed to read file metadata")
+                .is_file()
+        })
+        .for_each(|entry| {
+            let program =
+                fs::read_to_string(entry.path()).expect("failed to read quil program file");
+            let name = entry
+                .file_name()
+                .to_str()
+                .expect("bad filename")
+                .to_string();
+
```

Comment:
Perhaps it's the code golfer in me, but I feel like this `filter_map().filter().for_each()` could probably be collapsed.

Reply:
Fore!

Maybe it could? Would be happy to see if you have anything in mind. TBH it's good enough for me, and probably compiles down to the same thing a more visually compact version would. On the plus side, readability is still high enough here, where more combinator stuff might cut into that margin.